proposals/Simple-programs.md

```diff
@@ -0,0 +1,169 @@
+# Simple programs
+
+* [x] Proposed
+* [x] Prototype: Started
+* [ ] Implementation: Not Started
+* [ ] Specification: Not Started
+
+## Summary
+[summary]: #summary
+
+Allow a sequence of *statements* to occur right before the *namespace_member_declaration*s of a *compilation_unit* (i.e. source file).
+
+The semantics are that if such a sequence of *statements* is present, the following type declaration, modulo the actual type name and the method name, would be emitted:
+
+``` c#
+static class Program
+{
+    static async Task Main()
+    {
+        // statements
+    }
+}
+```
+
+See also https://github.com/dotnet/csharplang/issues/3117.
+
+## Motivation
+[motivation]: #motivation
+
+There's a certain amount of boilerplate surrounding even the simplest of programs,
+because of the need for an explicit `Main` method. This seems to get in the way of
+language learning and program clarity. The primary goal of the feature therefore is
+to allow C# programs without unnecessary boilerplate around them, for the sake of
+learners and the clarity of code.
+
+## Detailed design
+[design]: #detailed-design
+
+### Syntax
+
+The only additional syntax is allowing a sequence of *statement*s in a compilation unit,
+just before the *namespace_member_declaration*s:
+
+``` antlr
+compilation_unit
+    : extern_alias_directive* using_directive* global_attributes? statement* namespace_member_declaration*
```

Comment:
Should statements come before declarations or *instead of* them at this stage? What is the presumed process of migrating learners to "complex programs"?

Reply:
> Should statements come before declarations or instead of them at this stage? 

As the grammar indicates, top level statements should come before type/namespace declarations

---
In reply to: [375676793](https://github.com/dotnet/csharplang/pull/3159#discussion_r375676793) [](ancestors = 375676793)